[openapi] Add OpenAPI schema generation (#254)

TobiasWalle opened the pull request:
I added a plugin to generate OpenApi Definitions from the Javalin Code. Examples can be found in the Test.

## Api
### Configuration
To activate the plugin, the user has to enable the configuration
```kotlin
val app = Javalin.create { config ->
    config.enableOpenApi(
        Info().title("Example").version("1.0.0")
    )
}
```

The user can give an optional path, which create a GET route to the json.
```kotlin
val app = Javalin.create { config ->
    config.enableOpenApi(
        "docs/swagger.json",
        Info().title("Example").version("1.0.0")
    )
}
```

### Usage
There is now a higher order function, that allow to decorate a handler with metadata.
For example an api like this
```kotlin
app.get("/user") { ctx ->
    val name = ctx.queryParam("name")!!
    ctx.json(User(name = name))
}
```

can be documented like the following
```kotlin
app.get("/user", documented { ctx ->
    val name = ctx.queryParam("name")!!
    ctx.json(User(name = name))
}
        .queryParam<String>("name")
        .json<User>("200") {
            it.description = "Request successful"
        }
)
```

The api of the documentation matches the context api. To provide further information, a callback can be provided which can update the OpenApi metadata (See json response in example).

I added support for the complete Javalin Api.

### Access Schema directly
I also added a method on the javalin instance, which allows to get the schema directly. 
```kotlin
val mySchema = app.createOpenAPISchema()
```

tipsy commented:
Thanks @TobiasWalle, you work fast ! First things first, the build is failing because Travis is unstable, so don't worry about that. 

I'm looking at the example(s) you provided:

```
app.get("/user", documented { ctx ->
    val name = ctx.queryParam("name")!!
    ctx.json(User(name = name))
}
    .queryParam<String>("name")
    .json<User>("200") {
         it.description = "Request successful"
     }
)

// ------------------
app.get("/users", documented {
            val myCookie = it.cookie("my-cookie")
            val myHeader = it.cookie("x-my-header")
            val nameFilter = it.queryParam("name")
            val ageFilter = it.queryParam("age")
            it.json(listOf(User(name = "Jim")))
        }
                .cookie<String>("my-cookie") {
                    it.description = "My cookie"
                }
                .header<String>("x-my-header") {
                    it.description = "My header"
                }
                .pathParam<String>("my-path-param") {
                    it.description = "My path param"
                }
                .queryParam<String>("name") {
                    it.description = "The name of the users you want to filter"
                }
                .queryParam<Int>("age")
                .jsonArray<User>("200"))

// ------------------
app.get("/users2", documented { it.json(listOf(User(name = "Jim"))) }.json<Array<User>>("200"))
```

And the first thing I'm noticing is that there is pretty much a 1-1 mapping of Context calls and Swagger calls. This makes me wonder 🤔

It should be possible to build an abstraction layer to remove this duplication, but it would come at the price of not working with lambdas.

As I mentioned in #254, I'm not a big fan of the DSL approach in general, as it makes for very noisy controllers, and it's hard to see what's Javalin and what's Swagger.

Have you considered other approaches? Did consider (dare I say it) annotations?

TobiasWalle commented:
@tipsy Thank you! It was very fun to implement this :)

Also, thank you for your feedback. I didn't read the part in which you mentioned your dislike with the DSL approach. But I might can explain a little bit why I prefer it.

I agree that the DSL makes the controllers a lot more verbose. For simple examples it really is a 1:1 mapping of the implementation. For complex examples it is more like a short summary, but there is definitely duplication and the possibility to get things wrong.

Let's consider a (bit) more complex example.
```kotlin
app.put("/user", documented {
    val user = it.bodyAsClass(User::class.java)
    if (user.isInvalid()) {
        it.status(400).result("Bad Request")
    } else {
        it.status(204).result("Ok")
    }
}
        .body<User>()
        .result<String>("400")
        .result<String>("204")
)
```
In this example the distinction is a little more clearer, but there is still the potential for confusion.


I considered other approaches. As I read the issue again I want to summarise the findings. I think there are the following alternatives

1. Static code analysis (Parsing of comments and/or usage of methods)
2. Annotations with Reflection
3. Runtime Analysis

Before I start comparing the alternatives I want to explain the points that are important to me:
* **Bug resistance**. It should be hard to do something wrong. Silent errors or complex constructs outside the programming language should be avoided.
* **Avoid duplication**. We don't want to define everything twice.
* **Avoid verbosity**. We want to keep our controllers as clean as possible.
* **Avoid Javalin Api changes just for docs** We want do design our api as we prefer and still have the possiblity to add swagger documentation.
* **Avoid complexity** We want to have a very simple implementation to avoid bugs and make the software maintainable
* **Completeness** We want to make it possible to define all of the OpenApi fields

Let's check the alternatives.

## 1. Static code analysis
Let's take your example from the issue

```kotlin
/**
 * My very nice description
 *
 * @queryParam my-query-param some description
 * @pathParam user-id some other description
 * @result SerializeableObject object
 */
fun myHandler(ctx: Context) {
    val queryParam = ctx.queryParam("my-query-param")
    val userId = ctx.pathParam("user-id")
    ctx.json(SerializeableObject());
}

...

app.get("/my-path", Ctrl::myHandler)
```

We could write a program that parses the comments and/or the source the get the metadata

**Advantages**
Very easy to read. Our controllers just have comments now.

**Disadvantages**
We basically define a DSL in comments. This can get very complex, very quickly. I really want to avoid to write a custom parser, because it is very hard to get right.
While it works fine in the example above, we run into edge cases very quickly. For example we might want to add multiple responses for different status codes. We might also define the media types. Our comment might look now something like this:
```kotlin
/**
 * My very nice description
 *
 * @queryParam my-query-param some description
 * @pathParam user-id some other description
 * @result 200 [application/json, application/xml] SerializeableObject object
 * @result 500 [text/html] Error response
 */
```
We are still missing a lot of fields. We had to define a syntax for that.
This also doesn't solves the problem of duplication.
We might want to also scan the code to avoid duplication. But this would increase the complexity even more.

## 2. Annotations with Reflection
We could try to use annotations to define the metadata.

```kotlin
@Description("My very nice description")
@QueryParam("my-query-param", "some description")
@Result(SerializeableObject, "object")
fun myHandler(ctx: Context) {
    val queryParam = ctx.queryParam("my-query-param")
    val userId = ctx.pathParam("user-id")
    ctx.json(SerializeableObject());
}
```
In this case we avoid writing our own parser. My main concern is that we now design a DSL with annotations. And unfortunately we are very [limited](https://stackoverflow.com/a/2067863) in what we can do here. A complete OpenApi definition can be very complex and I think it can be hard to define without any runtime concepts.

## 3. Runtime Analysis
A problem is that we cannot access the results and parameters without running the handler. A possibility would be to log these things during runtime to access it later. The user wouldn't need to make sure that all the branches are executed (for example in his tests) and after that we could generate the schema. We would still need a DSL to define further metadata.


## A better DSL
Let's try to improve the DSL approach. I think the main problems are
* Duplication
* Verbosity
* Confusion with Javalin API

I think it is very hard to avoid the duplication issue. Because in Javalin, the handler definition is very dynamic (which I love) it is not easy to get this information without making sacrifices. 
While it's not perfect, I also don't think it's such a big problem in practice.

We can try to move the documentation away from the handler:

```kotlin
val myHandlerDocumentation = handlerDocumentation {
    description = "My very nice description"

    queryParam<String>("my-query-param") { description = "some description" }
    pathParam<String>("user-id") { description = "some other description" }
    json<SerializeableObject>("200") {  description = "object" }
}

val myHandler = documented(myHandlerDocumentation) { ctx ->
    val queryParam = ctx.queryParam("my-query-param")
    val userId = ctx.pathParam("user-id")
    ctx.json(SerializeableObject());
}
...

app.get("/my-path", myHandler)
```
I think this looks very similar to the comments approach. But it is completely typesafe and does not require any source code parsing.

What do you think about that?

If we can improve things with annotations, we can also give it a try. Maybe you can give me an example what you imagine?


Until tomorrow I won't have a lot of time for about a week. So don't wonder if I'm inactive during that time period.

tipsy commented:
Thanks for writing this all out @TobiasWalle - I'll go through each section and add my own pros/cons.

## 1. Static code analysis
I think my bias is showing when I say I like the idea of a comment based DSL. I personally don't use Swagger, so my immediate reaction is to want it "out of the way". I don't know how complex a Swagger spec gets, so it's probably a bad idea to build a comment parser, but let me add a couple of more pros:

* It's 100% language independent, will work equally well with Java, Kotlin, Scala, Groovy, etc
* You can put your comments anywhere (field, method, class)
* Not related to Javalin in any way, can be released as separate project

## 2. Annotations with Reflection

I have a branch locally for this, but since I'm not a Swagger user, it's hard for me to finish it. I'll show you what I have:

Implementation: 
```kotlin
annotation class SwaggerApi(
        val description: String,
        val path: String,
        val pathParams: Array<PathParam> = [],
        val queryParams: Array<QueryParam> = [],
        val formParams: Array<FormParam> = []
)

annotation class PathParam(val key: String, val value: String)
annotation class QueryParam(val key: String, val value: String)
annotation class FormParam(val key: String, val value: String)
```

Usage Kotlin:
```kotlin
@SwaggerApi(
        description = "My other nice description. It's just a single line.",
        path = "/users/:user-id",
        pathParams = [
            PathParam("user-id", "a very nice description"),
            PathParam("other-id", "another description")
        ],
        queryParams = [QueryParam("my-query-param", "some other description")],
        formParams = [FormParam("my-form-param", "a terrible description")]
)
```
Usage Java: 
```java
@SwaggerApi(
    description = "My other nice description. It's just a single line.",
    path = "/users/:user-id",
    pathParams = {
        @PathParam(key = "user-id", value = "a very nice description"),
        @PathParam(key = "other-id", value = "another description")
    },
    queryParams = {@QueryParam(key = "my-query-param", value = "some other description")},
    formParams = {@FormParam(key = "my-form-param", value = "a terrible description")}
)
```

Currently, I believe this is the best solution. It has all the pros of a comment based approach, but you don't have to deal with any custom parsing. You also get named params that work similarly (but not the same) in both Kotlin and Java. I'm not sure about other JVM languages, I'm guessing they all support the same features but with a different syntax.

* It's mostly language independent, should(!) work equally well with Java, Kotlin, Scala, Groovy, etc
* You can put your annotations anywhere (field, method, class)
* Not related to Javalin in any way, can be released as separate project
* Typesafe/has named params

## 3. Runtime Analysis

I don't think we should go down this road.

## A better DSL

Splitting up the documentation is definitely a huge improvement, but IMO it has some shortcomings compared to annotations/javadoc. There are a lot of different ways to declare a Javalin `Handler`. You can do fields/properties, methods/functions, classes, or you can use the Javalin `CrudHandler`, which is an interface with the 5 common crud operations (create, read all, read one, update, delete). Some people also create their own abstraction layer on top.

Going through the source/classes and looking for annotations is simple in all these cases, as they don't put any restraints on how you write your application. With the DSL you're forcing people to conform to a specific way of building apps, for each `Handler` you need a `documented` counter-part, and you can no longer use the `CrudHandler` or any other abstraction on top of `Handler` (correct me if I'm wrong).

The information your can extract from Javalin is very limited (verb, path), so I don't see the value in trying to integrate Swagger tightly (in code), unless you go in the opposite direction and create an abstraction layer which makes meta information available at compile time. I'm not a big fan of this idea (and it seems you aren't either), so please consider the annotation approach and let me know your thoughts.

TobiasWalle commented:
Thank you for your input. I see your point here and I agree that we should also support other interfaces (like the crud controller) which I totally forgot.

I will try to implement a small example that works with annotations. Maybe it is possible to do both (Annotations for simple controllers, DSL for more complex stuff)

You can expect the example in about 2 weeks, as I won't find the time next week.

I recommend you to check out the pet store swagger example as it uses many features of the documentation.

tipsy commented:
Great @TobiasWalle, looking forward to the update. I've looked at the petstore example, there doesn't seem to be anything in there that would be difficult to achieve with annotations, but I might be wrong. I'm actually not a big fan of annotations either, so I don't have much experience with them. This is one of the few cases where I think they make sense though.

TobiasWalle commented:
@tipsy I have some updates 😊

I updated my PR, so the API supports both, the annotations and an updated DSL. Internally the annotations will be converted to the DSL.

Let's look at the updated example:

**Annotations**
```kotlin
@OpenApi(
        description = "My very nice description",
        queryParams = [
            OpenApiParam("my-query-param", description = "some description")
        ],
        pathParams = [
            OpenApiParam("user-id", description = "some other description")
        ],
        responses = [
            OpenApiResponse(status = "200", returnType = SerializableObject::class)
        ]
)
fun myHandler(ctx: Context) {
    val queryParam = ctx.queryParam("my-query-param")
    val userId = ctx.pathParam("user-id")
    ctx.json(SerializeableObject());
}

app.get("/my-path", ::myHandler)
```

**DSL**
```kotlin
val myHandlerDocumentation = document()
        .operation { it.description = "My very nice description" }
        .queryParam<String>("my-query-param") { it.description = "some description" }
        .pathParam<String>("user-id") { it.description = "some other description " }
        .json<SerializableObject>("200")

fun myHandler(ctx: Context) {
    val queryParam = ctx.queryParam("my-query-param")
    val userId = ctx.pathParam("user-id")
    ctx.json(SerializeableObject());
}

app.get("/my-path", documented(myHandlerDocumentation, ::myHandler));
```

* Both APIs also support the crud handler (Take a look at the tests for an example).
* The Annotation API has the limitation that it doesn't work with inline or local functions. This is due to limitations in Kotlin. I added a warning, so the user know how to fix it, otherwise the annotation api will be ignored.
* If you wonder about the lambdas in the DSL Api. I have chosen this implementation, to avoid reimplementing the OpenApi Data Model. This allows me to support all the fields out of the box. In the annotation API I need to define all the fields again (Currently I added just the description to some fields).
* I refactored the path parsing to avoid the reimplementation of the parser for the documentation. Instead of string replacement, I now parse the given path string to a class representation and convert this representation back to the regex. Please check if you agree with my changes. Personally I think this implementation is also easier to extend.


I still have some work to do. A lot of fields are missing in the annotation API and I want to revisit my structure and the naming. I also want to add more tests, to cover more edge cases.

We also need to decide if we want to use the name "OpenApi" or "Swagger". "OpenApi" would be more correct, as it is the open standard that swagger uses.

What do you think about the changes so far?

tipsy commented:
@TobiasWalle this looks very good! `OpenApi` is probably a better name. The PR is getting a bit massive though, could you split it into four separate ones? 

* Formatting of exitsting code (in JavalinConfig you can move the `// @formatter:on` comment up to after `Inner`, I don't mind if the methods are properly formatted, but I want the field annotations to be on the same line).
* New path parser
* OpenApi DSL
* OpenApi annotations 

Great work, I'm excited that Javalin will finally support this!

TobiasWalle commented:
Glad you like it. I'm going to split my PR.

Just a question to the first point. Does the `// formatter:on` comment prevent the automatic formatting of Intellij? I tried to adapt my code style settings to the project.

tipsy commented:
Yes, that's what it does. I think it's an intellij specific syntax, but I'm not 100% sure.

TobiasWalle commented:
I reversed my formatting, so the first PR shouldn't be necessary.
I created a PR for the new parser https://github.com/tipsy/javalin/pull/575

TobiasWalle commented:
The PR of the core and the dsl is open now #576 

This is the vast majority of the code. The annotation api PR will be really small in comparison.
I'm looking forward for your feedback.

tipsy commented:
@TobiasWalle great work today! When you think you'll have the remaining PRs up?

TobiasWalle commented:
@tipsy Thank you! It is great that you process the PRs so fast. I think we are really productive.

I will now create the annotation PR. I think it will not take longer than an hour.

After that we can close this PR and I can start updating the swagger ui for the final PR (for now).

tipsy commented:
> I will now create the annotation PR. I think it will not take longer than an hour.

Great. I'll review today and hopefully release a new version tomorrow then. Either  `3.0.0.ALPHA1` or `3.0.0.RC0`, depending on the plugin PR you're also working on.

TobiasWalle commented:
Here it is #578

felipewom commented:
Is there any dependency need to inject OpenApi on javalin 3.0? 
I tried to 
`val app = Javalin.create { config ->
           config.enableOpenApi(
            OpenApiOptions(Info().version("1.0").description("My Application"))
                .path("/swagger-json")
                .swagger(SwaggerOptions("/swagger").title("My Swagger Documentation"))
                .reDoc(ReDocOptions("/redoc").title("My ReDoc Documentation"))
        )
...
`
And i got "Unresolved reference: enableOpenApi" or "Unresolved reference: Info"
thank you!

TobiasWalle commented:
@felipewom Yes you need to install at least:

<dependency>
    <groupId>io.swagger.core.v3</groupId>
    <artifactId>swagger-models</artifactId>
    <version>2.0.8</version>
</dependency>
<dependency>
    <groupId>cc.vileda</groupId>
    <artifactId>kotlin-openapi3-dsl</artifactId>
    <version>0.20.1</version>
</dependency>

For the Swagger UI you need 

<dependency>
    <groupId>org.webjars</groupId>
    <artifactId>swagger-ui</artifactId>
    <version>3.22.2</version>
</dependency>

I currently working on the documentation for this
